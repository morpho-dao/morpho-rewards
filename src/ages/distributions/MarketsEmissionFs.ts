export interface MarketsEmissionFs {
  epochId: string;
  totalEmission: string;
  snapshotProposal?: string;
  parameters: {
    snapshotBlock: number;
    initialTimestamp: number;
    finalTimestamp: number;
    duration: number;
  };
  markets: {
    [market: string]: {
      morphoEmittedSupplySide: string;
      morphoRatePerSecondSupplySide: string;
      morphoRatePerSecondBorrowSide: string;
      morphoEmittedBorrowSide: string;
      totalMarketSizeSupplySide: string;
      totalMarketSizeBorrowSide: string;
    };
  };
}
