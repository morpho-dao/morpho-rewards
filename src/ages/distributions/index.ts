export { ageOneDistribution } from "./ageOneDistribution";
export { ageTwoDistribution } from "./ageTwoDistribution";
export { ageThreeDistribution } from "./ageThreeDistribution";
export { ageFourDistribution } from "./ageFourDistribution/ageFourDistribution";
export { ageFiveDistribution } from "./ageFiveDistribution/ageFiveDistribution";
export { ageSixDistribution } from "./ageSixDistribution/ageSixDistribution";
