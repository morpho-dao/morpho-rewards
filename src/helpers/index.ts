export * from "./time";
export * from "./constants";
export * from "./maths";
export * from "./parsers";
