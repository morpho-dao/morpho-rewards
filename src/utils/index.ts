export { userBalancesToUnclaimedTokens, getUserRewards, MarketRewards, sumRewards } from "./getUserRewards";
export { getAccumulatedEmission, getAccumulatedEmissionPerMarket } from "./accumulatedEmission";
export { getUserBalances } from "./getUserBalances";
export * from "./markets";
export * from "./merkleTree";
export * from "./graph";
export * from "./chain";
export * from "./etherscan";
