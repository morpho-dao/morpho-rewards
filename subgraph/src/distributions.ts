import { Address, BigInt, log } from "@graphprotocol/graph-ts";
import { emissions, epochToEndTimestamps, epochToStartTimestamps } from "./generated-emissions";


export const epochNumberToStartTimestamp = (epochNumber: i32): BigInt | null => {

  const ts = epochToStartTimestamps.get(`epoch-${epochNumber}`);
  if (!ts) return null;
  return BigInt.fromString(ts.toString());
};
export const epochNumberToEndTimestamp = (epochNumber: i32): BigInt | null => {
  const ts = epochToEndTimestamps.get(`epoch-${epochNumber}`);
  if (!ts) return null;
  return BigInt.fromString(ts.toString());
};
export const timestampToEpochId = (timestamp: BigInt): i32 => {
  if (timestamp.le(BigInt.fromI32(1654707606))) return 0;
  let epoch = 1;
  while (true) {
    let endTimestamp = epochNumberToEndTimestamp(epoch);
    if (!endTimestamp) return epoch;
    if (timestamp.le(endTimestamp)) return epoch;
    epoch++;
  }
};
export const fetchDistribution = (
  timestamp: BigInt,
  side: string,
  market: Address
): BigInt => {
  if (timestamp.lt(BigInt.fromI32(1654707606))) return BigInt.zero();
  const epochNumber = timestampToEpochId(timestamp);
  if (!epochNumber) {
    log.debug("No epoch id at ts {}", [timestamp.toString()]);
    return BigInt.zero();
  }
  const id = `epoch-${epochNumber}-${side}-${market.toHexString()}`;
  return fetchDistributionFromDistributionId(id);
};

export const fetchDistributionFromDistributionId = (id: string): BigInt => {
  const emission = emissions.get(id);
  if (!emission) return BigInt.zero();
  return emission;
};
