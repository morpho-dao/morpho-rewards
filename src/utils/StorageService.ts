import * as path from "path";
import * as fs from "fs";
import { MarketsEmission } from "../ages/distributions/MarketsEmission";
import { Proof, Proofs } from "../ages/distributions/Proofs";
import { numberOfEpochs } from "../ages/ages";
import { UsersDistribution } from "../ages/distributions/UsersDistribution";

export interface StorageService {
  readMarketDistribution: (age: string, epoch: string) => Promise<MarketsEmission | void>;
  writeMarketEmission: (age: string, epoch: string, emission: MarketsEmission) => Promise<void>;
  readUsersDistribution: (age: string, epoch: string) => Promise<UsersDistribution | void>;
  writeUsersDistribution: (age: string, epoch: string, distribution: UsersDistribution) => Promise<void>;
  readProofs: (epoch: number) => Promise<Proofs | void>;
  readAllProofs: () => Promise<Proofs[]>;
  readUserProof: (epoch: number, address: string) => Promise<Proof | void>;
  writeProofs: (epoch: number, proofs: Proofs) => Promise<void>;
}

export type ProofsCache = { [epoch: number]: Proofs | undefined };
export type MarketsEmissionCache = {
  [age: string]: {
    [epoch: string]: MarketsEmission | undefined;
  };
};
export type UsersDistributionCache = {
  [age: string]: {
    [epoch: string]: UsersDistribution | undefined;
  };
};

export class FileSystemStorageService implements StorageService {
  #emissionsCache: MarketsEmissionCache = {};
  #distributionsCache: UsersDistributionCache = {};
  #proofsCache: ProofsCache = {};
  #distributionRoot = "../../distribution";

  #generateDistributionPath(age: string, epoch: string) {
    const folder = path.resolve(__dirname, this.#distributionRoot, age, epoch);
    const file = path.resolve(folder, "marketsEmission.json");
    return { folder, file };
  }

  #generateProofsPath(epoch: number) {
    const folder = path.resolve(__dirname, this.#distributionRoot, "proofs");
    const filename = `proofs-${epoch}.json`;
    const file = path.resolve(folder, filename);
    return { folder, file };
  }

  #generateUsersDistributionPath(age: string, epoch: string) {
    const folder = path.resolve(__dirname, this.#distributionRoot, age, epoch);
    const file = path.resolve(folder, "usersDistribution.json");
    return { folder, file };
  }

  async readMarketDistribution(age: string, epoch: string) {
    try {
      const inCache = this.#emissionsCache[age]?.[epoch];
      if (inCache) return inCache;
      const { file } = this.#generateDistributionPath(age, epoch);
      const distribution = require(file) as MarketsEmission;
      this.#emissionsCache[age] = { ...this.#emissionsCache[age], [epoch]: distribution };
      return distribution;
    } catch (error) {
      return;
    }
  }

  async writeMarketEmission(age: string, epoch: string, emission: MarketsEmission) {
    const { folder, file } = this.#generateDistributionPath(age, epoch);
    const fileExists = await fs.promises
      .access(file, fs.constants.R_OK | fs.constants.W_OK)
      .then(() => true)
      .catch(() => false);
    if (fileExists) throw new Error(`File ${file} already exists, can't write it.`);
    await fs.promises.mkdir(folder, { recursive: true });
    await fs.promises.writeFile(file, JSON.stringify(emission, null, 2));
  }

  async readUsersDistribution(age: string, epoch: string) {
    try {
      const inCache = this.#distributionsCache[age]?.[epoch];
      if (inCache) return inCache;
      const { file } = this.#generateUsersDistributionPath(age, epoch);
      const distribution = require(file) as UsersDistribution;
      this.#distributionsCache[age] = { ...this.#distributionsCache[age], [epoch]: distribution };
      return distribution;
    } catch (error) {
      return;
    }
  }

  async writeUsersDistribution(age: string, epoch: string, distribution: UsersDistribution) {
    const { folder, file } = this.#generateUsersDistributionPath(age, epoch);
    const fileExists = await fs.promises
      .access(file, fs.constants.R_OK | fs.constants.W_OK)
      .then(() => true)
      .catch(() => false);
    if (fileExists) throw new Error(`File ${file} already exists, can't write it.`);
    await fs.promises.mkdir(folder, { recursive: true });
    await fs.promises.writeFile(file, JSON.stringify(distribution, null, 2));
  }

  async readProofs(epoch: number) {
    try {
      const inCache = this.#proofsCache[epoch];
      if (inCache) return inCache;
      const { file } = this.#generateProofsPath(epoch);
      const proofs = require(file) as Proofs;
      this.#proofsCache[epoch] = proofs;
      return proofs;
    } catch (error) {
      return;
    }
  }

  async readAllProofs() {
    const result = await Promise.all(
      new Array(numberOfEpochs).fill(0).map(async (_, index) => {
        const epoch = numberOfEpochs - index;
        const proofs = await this.readProofs(epoch);
        return proofs ? [proofs] : [];
      })
    );
    return result.flat();
  }

  async readUserProof(epoch: number, address: string) {
    const proof = await this.readProofs(epoch);
    return proof?.proofs?.[address];
  }

  async writeProofs(epoch: number, proofs: Proofs) {
    const { folder, file } = this.#generateProofsPath(epoch);
    const fileExists = await fs.promises
      .access(file, fs.constants.R_OK | fs.constants.W_OK)
      .then(() => true)
      .catch(() => false);
    if (fileExists) throw new Error(`File ${file} already exists, can't write it.`);
    await fs.promises.mkdir(folder, { recursive: true });
    await fs.promises.writeFile(file, JSON.stringify(proofs, null, 2));
  }
}
