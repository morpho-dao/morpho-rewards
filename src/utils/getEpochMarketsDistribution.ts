import { providers } from "ethers";
import { ages } from "../ages";
import { timestampToEpoch } from "./timestampToEpoch";
import { MarketsEmission } from "../ages/distributions/MarketsEmission";
import { formatUnits } from "ethers/lib/utils";
import { now } from "../helpers";
import { StorageService } from "./StorageService";

export const getMarketsDistribution = async (
  storageService: StorageService,
  timestamp?: number,
  provider: providers.Provider = new providers.InfuraProvider(1)
) => {
  const epochConfig = timestampToEpoch(timestamp ?? now());
  if (!epochConfig) throw Error(`No epoch found at timestamp ${timestamp}`);
  return getEpochMarketsDistribution(epochConfig.epoch.id, provider, storageService);
};

export const getEpochMarketsDistribution = async (
  epochId: string,
  provider: providers.Provider,
  storageService: StorageService
) => {
  const [age, epoch] = epochId.split("-");
  const distribution = await storageService.readMarketDistribution(age, epoch);
  if (distribution) return distribution;
  // need to compute distribution from chain
  return computeEpochMarketsDistribution(age, epoch, provider, storageService);
};

export const computeEpochMarketsDistribution = async (
  age: string,
  epoch: string,
  provider: providers.Provider,
  storageService: StorageService
) => {
  const distribution = await storageService.readMarketDistribution(age, epoch);
  if (distribution) return distribution;
  console.warn(`Commpute distribution for epoch ${age}-${epoch}`);
  const ageConfig = ages.find((a) => a.ageName === age);
  if (!ageConfig) throw Error(`Unknown age: ${age}`);
  const epochConfig = ageConfig.epochs.find((e) => e.epochName === epoch);
  if (!epochConfig) throw Error(`Unknown epoch: ${age}-${epoch}`);

  // Will revert if snapshotBlock is not defined
  const { marketsEmissions } = await ageConfig.distribution(ageConfig, epochConfig, provider);

  const formattedMarketsEmissions = Object.fromEntries(
    Object.entries(marketsEmissions).map(([key, marketConfig]) => [
      key,
      {
        supply: formatUnits(marketConfig!.supply),
        supplyRate: marketConfig!.supplyRate.toString(),
        borrowRate: marketConfig!.borrowRate.toString(),
        borrow: formatUnits(marketConfig!.borrow),
      },
    ])
  );
  const result: MarketsEmission = {
    age,
    epoch,
    totalEmission: epochConfig.totalEmission.toString(),
    parameters: {
      snapshotBlock: epochConfig.snapshotBlock!.toString(),
      initialTimestamp: epochConfig.initialTimestamp.toString(),
      finalTimestamp: epochConfig.finalTimestamp.toString(),
      duration: epochConfig.finalTimestamp.sub(epochConfig.initialTimestamp).toString(),
    },
    markets: formattedMarketsEmissions,
  };
  await storageService.writeMarketEmission(age, epoch, result);
  return result;
};
