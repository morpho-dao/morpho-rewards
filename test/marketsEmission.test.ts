import { ages } from "../src";
import { getMarketsDistribution } from "../src/utils/getEpochMarketsDistribution";
import { getDefaultProvider, providers } from "ethers";
import { MarketEmission } from "../src/utils";
import { Optional } from "../src/helpers/types";
import { aStEth, cFei } from "../src/helpers";
import { FileSystemStorageService } from "../src/utils/StorageService";
import { MarketsEmissionFs } from "../src/ages/distributions/MarketsEmissionFs";
import { formatUnits } from "ethers/lib/utils";
import {
  getAaveMarketsParameters,
  getAaveV3MarketsParameters,
  getCompoundMarketsParameters,
} from "../src/utils/markets/fetchMarketsData";

const storageService = new FileSystemStorageService();

describe("Test Market data fetching", () => {
  const provider = getDefaultProvider(process.env.RPC_URL);
  const blockTag = 17_000_000;
  describe("Compound", () => {
    it("Should fetch Compound data", async () => {
      const marketParamers = await getCompoundMarketsParameters(blockTag, provider);
      expect(marketParamers).toMatchSnapshot();
    });
  });
  describe("Aave", () => {
    it("Should fetch Aave v2 data", async () => {
      const marketParamers = await getAaveMarketsParameters(blockTag, provider);
      expect(marketParamers).toMatchSnapshot();
    });
    it("Should fetch Aave v3 data", async () => {
      const MA3_DEPLOYMENT_BLOCK = 17161283;
      const marketParamers = await getAaveV3MarketsParameters(MA3_DEPLOYMENT_BLOCK + 1, provider);
      expect(marketParamers).toMatchSnapshot();
    });
  });
});

describe.each(ages)("Test Ages Distributions", (age) => {
  const EPOCHS_PER_AGE = ages.indexOf(age) < 3 ? 3 : 1;
  const ts = Math.floor(Date.now() / 1000);
  const provider = new providers.JsonRpcProvider(process.env.RPC_URL);
  it(`Should have the correct configuration for ${age.ageName}`, () => {
    expect(age.ageName).not.toBeUndefined();
    expect(age.epochs).toHaveLength(EPOCHS_PER_AGE);
    expect(age.startTimestamp).not.toBeUndefined();
    expect(age.endTimestamp).not.toBeUndefined();
  });
  it(`Should have consistent timestamps for ${age.ageName}`, () => {
    const firstEpoch = age.epochs[0];
    const lastEpoch = age.epochs[age.epochs.length - 1];
    expect(age.startTimestamp).toBnEq(firstEpoch.initialTimestamp);
    expect(age.endTimestamp).toBnEq(lastEpoch.finalTimestamp);
  });

  if (age.startTimestamp.lt(ts)) {
    it(`Should return a market configuration for ${age.ageName}`, async () => {
      const marketEmission = await getMarketsDistribution(storageService, age.startTimestamp.toNumber());
      expect(marketEmission).not.toBeUndefined();
    });
  }

  describe.each(age.epochs)(`Test each epochs of ${age.ageName}`, (epochConfig) => {
    const { initialTimestamp, finalTimestamp, initialBlock, finalBlock, snapshotBlock, epochNumber } = epochConfig;

    if (initialTimestamp.lt(ts)) {
      it(`Should have a snapshot block for epoch ${epochNumber}`, () => {
        expect(snapshotBlock).not.toBeUndefined();
      });
      it(`Should have an initial block for epoch ${epochNumber}`, () => {
        expect(initialBlock).not.toBeUndefined();
      });

      it(`Should have a distribution computed for epoch ${epochNumber}`, async () => {
        let file: object | undefined = undefined;
        try {
          file = (await storageService.readMarketDistribution(epochNumber))!;
        } catch (e) {
          console.error(e);
        }
        expect(file).not.toBeUndefined();
      });

      describe(`Distribution for epoch ${epochNumber}`, () => {
        let marketsEmissions: { [p: string]: Optional<MarketEmission> } = {};
        let file: MarketsEmissionFs;
        beforeAll(async () => {
          file = (await storageService.readMarketDistribution(epochNumber))!;
          ({ marketsEmissions } = await age.distribution(age, epochConfig, provider));
        });
        it(`Should have a correct number of markets computed for epoch ${epochNumber}`, async () => {
          expect(Object.values(marketsEmissions).length).toEqual(Object.values(file.markets).length);
        });
        it(`Should have the correct markets computed for epoch ${epochNumber}`, async () => {
          expect(Object.keys(marketsEmissions)).toEqual(Object.keys(file.markets));
        });
        it(`Should have the correct supply rates computed for epoch ${epochNumber}`, async () => {
          Object.entries(marketsEmissions).forEach(([market, emission]) =>
            expect(formatUnits(emission!.morphoRatePerSecondSupplySide)).toEqual(
              file.markets[market].morphoRatePerSecondSupplySide
            )
          );
        });
        it(`Should have the correct borrow rates computed for epoch ${epochNumber}`, async () => {
          Object.entries(marketsEmissions).forEach(([market, emission]) =>
            expect(formatUnits(emission!.morphoRatePerSecondBorrowSide)).toEqual(
              file.markets[market].morphoRatePerSecondBorrowSide
            )
          );
        });
        if (epochNumber >= 4) {
          it(`Should not distribute tokens on Compound FEI deprecated market for epoch ${epochNumber}`, async () => {
            expect(marketsEmissions[cFei]).toBeUndefined(); // no distribution for the fei token
          });

          it(`Should not distribute tokens to StEth borrowers on Aave for epoch ${epochNumber}`, async () => {
            expect(marketsEmissions[aStEth]?.morphoRatePerSecondBorrowSide.isZero()).toEqual(true); // no borrowers of steth on aave
            expect(marketsEmissions[aStEth]?.morphoEmittedBorrowSide.isZero()).toEqual(true); // no borrowers of steth on aave
          });
        }
      });
    }
    if (finalTimestamp.lt(ts)) {
      it(`Should have a final block for epoch ${epochNumber}`, () => {
        expect(finalBlock).not.toBeUndefined();
      });
    }
  });
});
