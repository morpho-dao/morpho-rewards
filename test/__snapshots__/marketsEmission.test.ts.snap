// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Market data fetching Aave Should fetch Aave v2 data 1`] = `
[
  {
    "address": "0x028171bCA77440897B824Ca71D1c56caC55b68A3",
    "decimals": 18,
    "morphoBorrowMarketSize": {
      "hex": "0x070f308ea664908b45889e",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x0c8899f7fed78085d6c810",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x0c4e",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x01ea2bd6b53d75",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x0f4a75f90183ef5a62a2",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x23e442bc50a6b00dbb73",
      "type": "BigNumber",
    },
  },
  {
    "address": "0x030bA81f1c18d280636F32af80b9AAd02Cf0854e",
    "decimals": 18,
    "morphoBorrowMarketSize": {
      "hex": "0x1523895829aff5c80ed3",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x10d7d72af5f9308cc140",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x1284",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x0de0b6b3a7640000",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x4895d95b52acbcba0f36",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x8d4836de4d58470f6f1d",
      "type": "BigNumber",
    },
  },
  {
    "address": "0xBcca60bB61934080951369a648Fb03DF4F96263C",
    "decimals": 6,
    "morphoBorrowMarketSize": {
      "hex": "0x110ab7316378",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x37b0dad82def",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x1036",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x01e61e2109ede3",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x30a9bb14da08b5f44668",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x4788e68fa14d48fe2629",
      "type": "BigNumber",
    },
  },
  {
    "address": "0x3Ed3B47Dd13EC9a98b44e6204A523E766B225811",
    "decimals": 6,
    "morphoBorrowMarketSize": {
      "hex": "0x135d0c5cc31d",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x0be9be175671",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x1bb2",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x01ea41ff223f5a",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x22d9b6b93b9053149018",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x3d3f8b11d1b3e2622d66",
      "type": "BigNumber",
    },
  },
  {
    "address": "0x9ff58f4fFB29fA2266Ab25e75e2A8b3503311656",
    "decimals": 8,
    "morphoBorrowMarketSize": {
      "hex": "0x0f6f404c0d",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x2b3fb1d179",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x11bc",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0xd0cf579448054e00",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x0ac545bdea0aff234d03",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x58b65cdb4988085a333d",
      "type": "BigNumber",
    },
  },
  {
    "address": "0x1982b2F5814301d4e9a8b0201555376e62F82428",
    "decimals": 18,
    "morphoBorrowMarketSize": {
      "hex": "0x00",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x1a5a17ff3e1298b722a4",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x00",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x0dd42a81d4e12258",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x00",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0xbec16c68e1e0f4c59edc",
      "type": "BigNumber",
    },
  },
  {
    "address": "0x8dAE6Cb04688C62d939ed9B68d32Bc62e49970b1",
    "decimals": 18,
    "morphoBorrowMarketSize": {
      "hex": "0xc64365b8594b38cc276a",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x0177a8eb4a7e9f4b364313",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x1388",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x01f1b26300a0ef",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x0471c438bddc917826d3",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x18ef255389bd2952105f",
      "type": "BigNumber",
    },
  },
]
`;

exports[`Test Market data fetching Aave Should fetch Aave v3 data 1`] = `[]`;

exports[`Test Market data fetching Compound Should fetch Compound data 1`] = `
[
  {
    "address": "0x5d3a536e4d6dbd6114cc1ead35777bab948e3643",
    "decimals": 18,
    "morphoBorrowMarketSize": {
      "hex": "0x04bb493fb9c7ce3aa876a9",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x0a7daf707c624f67cd5816",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x0d05",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x0de012fe2354c000",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0xa19c460f4b1608f6dc0c4b",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x0133b74469a355b9e75ae7a8",
      "type": "BigNumber",
    },
  },
  {
    "address": "0x70e36f6bf80a52b3b46b3af8e106cc0ed743e8e4",
    "decimals": 18,
    "morphoBorrowMarketSize": {
      "hex": "0x0113caea74b083ec0f1d",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x011bfcd8f1e7d3b583ac",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x1715",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x025076135b2b9f3000",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x752f3b18ce18eb1a9542",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x06a0976ff54cca455a955b",
      "type": "BigNumber",
    },
  },
  {
    "address": "0x4ddc2d193948926d02f9b1fe9e1daa0718270ed5",
    "decimals": 18,
    "morphoBorrowMarketSize": {
      "hex": "0x539611ed75d70fd529",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0xfdfbe9e361beb60108",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x0320",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x6525837a31b96c2000",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x0e40b31fd86e872c51c54e",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x01c31d43eb2ed1026659f323",
      "type": "BigNumber",
    },
  },
  {
    "address": "0x35a18000230da775cac24873d00ff85bccded550",
    "decimals": 18,
    "morphoBorrowMarketSize": {
      "hex": "0x10e83f306ca2f983ba84",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x222bd0cb3e91f8b31d53",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x0bc9",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x537c6cf495bab000",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x02f6c4fc310e9048d93ea6",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x16c7c5b0a910c6410bb1bf",
      "type": "BigNumber",
    },
  },
  {
    "address": "0x39aa39c021dfbae8fac545936693ac917d5e7563",
    "decimals": 6,
    "morphoBorrowMarketSize": {
      "hex": "0x044e0809313a",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x0da8cad839bb",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x0bca",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x0c9f2c9cd04674edea40000000",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0xb5fa9e0eaa4f201ff69000",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x0189c08ffa2d7a2e221fe000",
      "type": "BigNumber",
    },
  },
  {
    "address": "0xf650c3d88d12db855b8bf7d11be6c55a4e07dcc9",
    "decimals": 6,
    "morphoBorrowMarketSize": {
      "hex": "0x041d8e1883e8",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x038ed149face",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x16d0",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x0c9f2c9cd04674edea40000000",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x686193b7ed304a7f2f1000",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x9fb7f2ce9d812e2a5fb000",
      "type": "BigNumber",
    },
  },
  {
    "address": "0xccf4429db6322d5c611ee964527d42e5d685dd6a",
    "decimals": 8,
    "morphoBorrowMarketSize": {
      "hex": "0x01b30c42c3",
      "type": "BigNumber",
    },
    "morphoSupplyMarketSize": {
      "hex": "0x052e4a4eaf",
      "type": "BigNumber",
    },
    "p2pIndexCursor": {
      "hex": "0x0d05",
      "type": "BigNumber",
    },
    "price": {
      "hex": "0x0dc4175821ad4673d6695f000000",
      "type": "BigNumber",
    },
    "totalPoolBorrowUSD": {
      "hex": "0x02ec69dabacb352db24780",
      "type": "BigNumber",
    },
    "totalPoolSupplyUSD": {
      "hex": "0x0136a0cc23917a791948ee80",
      "type": "BigNumber",
    },
  },
]
`;
